driver: sensor: Add sht21 driver

aakkds opened the pull request:
Add htu21d humidity and temperature sensor driver and a sample using it on the hexiwear board.

EDIT: Changed to sht21 driver because htu21d (and si7021) are clones, as pointed by @pabigot

aakkds committed:
driver: sensor: Add driver for sht21 temp and humidity sensor

This driver also work with htu21d and si7021 clones.

Signed-off-by: Philémon Jaermann <p.jaermann@gmail.com>

aakkds committed:
boards: arm: Enable htu21d for hexiwear board

Htu21d is present on the hexiwear board

Signed-off-by: Philémon Jaermann <p.jaermann@gmail.com>

aakkds committed:
samples: sensor: Add sample for htu21d sensor

Test sample using hexiwear board

Signed-off-by: Philémon Jaermann <p.jaermann@gmail.com>

aakkds committed:
tests: drivers: Enable build test for sht21 driver

Set dts_fixup for test build

Signed-off-by: Philémon Jaermann <p.jaermann@gmail.com>

pabigot reviewed (changes requested):
I haven't looked at the technical aspects of this yet, but I do have a naming change I would like to have considered:

The Measurement Specialities HTU21D is a clone of the Sensirion SHT21; so is the Silicon Labs Si7021.  The same driver can support all three devices (I have [one for a different sensor API](https://github.com/pabigot/zephyr/commit/3f1d74a4622a321b2aaa4ca7ade942d649d3cd77).

As the SHT21 (`sensirion,sht21`) is the original, I would like to see the driver be based on that name, as it is most likely to be recognized.

Is that acceptable?  There should be no need to change the implementation much beyond changing HTI21D to SHT21 everywhere.  I have all three variants and can verify it works on all of them.  (The only behavior variation between them I'm aware of is the delay required before the measurement becomes available.)

aakkds commented:
@pabigot Thank you very much for your input! I was not aware of this, but what you suggest sounds like the way to go.
Your comment also reminded me that the current implementation uses the "holds" command and I should switch to the "no hold" command to have the bus available during measurement time.

I do not have a sht21 nor a si7021 so that would be very welcome if you can validate once I make the change (most likely toward the end of week/  next week end).

Cheers

aakkds committed:
drivers: sensor: Add Si7021 specific heater current consumption

The current consumption of the on chip heater, influencing
the heating effect, is only available on the Si7021 sensor.

Signed-off-by: Philémon Jaermann <p.jaermann@gmail.com>

aakkds commented:
@pabigot I implemented the changes. Except the wait time that are different, there is also the possibility to control the current of the on chip heating element that is possible on the Si7021 chip only.
That would be great if you can test this using a SHT21 and a Si7021.
Let me know if this corresponds or not to what you had in mind.

pabigot commented on drivers/sensor/sht21/Kconfig:
```diff
@@ -0,0 +1,55 @@
+config SHT21_ENABLE_OTP_RELOAD
+	bool "Enable OTP reload"
+
+if SHT21_ENABLE_HEATER
```
I think this and other heater stuff should be removed.

The purpose of the heater is for testing the sensor, generally on startup: enable the heater and see if the temperature increases, disable it and it should drop.  I can't imagine a reasonable use case for turning it on at startup and leaving it on.  There's also no API for turning it on and off at runtime, though it could be added through [specific extensions](https://docs.zephyrproject.org/latest/reference/drivers/index.html#device-specific-api-extensions).

pabigot commented on drivers/sensor/sht21/sht21.c:
```diff
@@ -0,0 +1,203 @@
+
+	user_cfg |= (SHT21_RH_RESOLUTION |
+		     SHT21_TEMP_RESOLUTION |
+		     SHT21_HEATER |
```
See comment on Kconfig.  There's no reason the heater should be enabled on startup and left that way.

pabigot commented on drivers/sensor/sht21/sht21.h:
```diff
@@ -0,0 +1,108 @@
+#define SHT21_MEAS_RH_WAIT_TIME		(16)
+#define SHT21_MEAS_TEMP_WAIT_TIME	(50)
+#elif DT_INST_0_SENSIRION_SHT21_VARIANT_SI7021
+#define SHT21_MEAS_RH_WAIT_TIME		(12)
```
FYI: For the one Si7021 I have to hand, 12 ms is not long enough (although the datasheet does specify it should be).  I needed to increase this to 18 to get a reliable read.

There's a risk that other durations for Si7021 and other variants are too tight.

pabigot commented on samples/sensor/htu21d/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2020, Philémon Jaermann
```
The sample should be named `samples/sensor/sht21` since that's the name of the driver, even if the in-tree example uses an HTU21D.

pabigot commented on drivers/sensor/sht21/Kconfig:
```diff
@@ -0,0 +1,55 @@
+	help
+	  Specify the desired measurement resolution.
+
+config SHT21_MEAS_RES_RH_8_BITS
```
As a maintainer looking at the implementation: it would be nice if the symbol name encoded both the RH and T resolutions, because they jump around.  Perhaps 12RH_14T (0), 8RH_12T (1), 10RH_13T (2), and 11RT_11T (3).

pabigot reviewed (changes requested):
Sorry for the delay in reviewing.  On the whole this is good, just a few comments:

* IMO the heater support should be removed as it's not something normally used and there's no API to use it properly, see discussion in comments.
* The sample name should match the driver name;
* The timing is too tight for at least one Si7021 resolution.  (I've tested the default resolution on HTU21D, SHT21, and Si7021.  I haven't tested any other resolutions.  I may do that next review.)

The conditional checking in the header to determine the right delays for a specific instance and resolution is horrific, but probably not worth refactoring until we have support for multiple instances.  It could be dramatically simplified by just assuming the Sensirion limits are used for all variants, and figuring out how to relax them for other devices later.

panicking commented:
I will rebase test it on my hardware and check if it's working as expected @str4t0m

str4t0m commented:
> I will rebase test it on my hardware and check if it's working as expected @str4t0m

I have linked that comment above, because it was the first time this discussion came up in zephyr, I assume you will try the other open PR https://github.com/zephyrproject-rtos/zephyr/pull/56227, or won't you?

panicking commented:
@str4t0m yes that one. I need only to prepare a new driver upstream. Hope in the next ;)
